src/vsr/replica.zig

```diff
@@ -2923,26 +3005,65 @@ pub fn ReplicaType(
         ///
         /// The caller owns the returned message, if any, which has exactly 1 reference.
         fn create_view_change_message(self: *Self, command: Command) *Message {
-            // We may send a start_view message in normal status to resolve a backup's view jump:
             assert(self.status == .normal or self.status == .view_change);
             assert((self.status == .normal) == (command == .start_view));
             assert((self.status == .view_change) == (command == .do_view_change));
-            assert(self.view >= self.log_view);
             assert(self.view >= self.view_durable());
             assert(self.log_view >= self.log_view_durable());
-
-            assert(command != .do_view_change or self.log_view < self.view);
-            assert(command != .start_view or self.log_view == self.view);
+            assert((self.log_view < self.view) == (command == .do_view_change));
+            assert((self.log_view == self.view) == (command == .start_view));
+            assert(self.log_view < self.view or self.replica == self.primary_index(self.view));
 
             const message = self.message_bus.get_message();
             defer self.message_bus.unref(message);
 
-            const headers = self.create_view_change_headers();
-            assert(headers.array.len > 0);
-            assert(headers.array.get(0).op == self.op);
+            switch (command) {
+                // The DVC headers are already up to date, either via:
+                // - transition_to_view_change_status(), or
+                // - superblock's vsr_headers (after recovery).
+                .do_view_change => assert(self.view_headers.array.get(0).op >= self.op),
+                .start_view => {
+                    self.view_headers.array.len = 0;
+
+                    var op = self.op + 1;
+                    while (op > 0 and
+                        self.view_headers.array.len < constants.view_change_headers_suffix_max)
+                    {
+                        op -= 1;
+                        self.view_headers.array.appendAssumeCapacity(self.journal.header_with_op(op).?.*);
+                    }
+                    assert(self.view_headers.array.len + 2 <= constants.view_change_headers_max);
+
+                    // Determine the consecutive extent of the log that we can help recover.
+                    // This may precede op_repair_min if we haven't had a view-change recently.
+                    const range_min = (self.op + 1) -| constants.journal_slot_count;
+                    const range = self.journal.find_latest_headers_break_between(range_min, self.op);
+                    const op_min = if (range) |r| r.op_max + 1 else range_min;
+                    assert(op_min <= op);
+                    assert(op_min <= self.op_repair_min());
```

Comment:
Hm, this is interesting -- we send out SV, but we actually don't persist that info anywhere. So backups might learn new headers without learning new prepares. I think this means that we need CTRL for liveness even if we don't have disk faults, something I didn't realize before. 

The situation I am thinking is that a primary accepts a request, increments its op, but the prepare subesquently gets dropped by the network, and doesn't make it to disk. Now, if the primary sends out SVs which inlcude it's current `op` and then dies, we are in the situation where:

* the cluster as a whole knows the `op`
* no data file contains a `prepare` for this `op`
* no disk faults occurred.